rmf_fleet_adapter/src/rmf_fleet_adapter/events/internal_ChopeNegotiator.hpp

```diff
@@ -0,0 +1,350 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef SRC__RMF_FLEET_ADAPTER__AGV__INTERNAL_CHOPE_NEGOTIATOR_HPP
+#define SRC__RMF_FLEET_ADAPTER__AGV__INTERNAL_CHOPE_NEGOTIATOR_HPP
+
+#include "../agv/RobotContext.hpp"
+
+#include <memory.h>
+
+namespace rmf_fleet_adapter {
+namespace chope {
+
+/// This class implements the protocol for negotiating a spot with the "chope"
+/// node. The chope node maintains a list of spots which are free.
+class ChopeNodeNegotiator:
+  public std::enable_shared_from_this<ChopeNodeNegotiator>
+{
+public:
+  static std::shared_ptr<ChopeNodeNegotiator> make(
+    std::shared_ptr<agv::RobotContext> context,
+    const std::vector<rmf_traffic::agv::Plan::Goal> goals,
+    const bool same_map,
+    const std::function<void(const rmf_traffic::agv::Plan::Goal&)>
+      selected_final_destination_cb,
+    const std::function<void(const rmf_traffic::agv::Plan::Goal&)> selected_waitpoint_cb)
+  {
+
+    auto negotiator = std::make_shared<ChopeNodeNegotiator>();
+    negotiator->_context = context;
+    negotiator->_goals = std::move(goals);
+    negotiator->_selected_final_destination_cb = std::move(selected_final_destination_cb);
+    negotiator->_selected_waitpoint_cb = std::move(selected_waitpoint_cb);
+    negotiator->_reservation_id = negotiator->_context->last_reservation_request_id();
+    negotiator->_reservation_ticket =
+      negotiator->_context->node()->location_ticket_obs().observe_on(rxcpp::identity_same_worker(
+        negotiator->_context->worker()))
+      .subscribe([w =
+        negotiator->weak_from_this()](const std::shared_ptr<rmf_chope_msgs::msg::Ticket>
+        &msg)
+      {
+
+        const auto self = w.lock();
+        if (!self)
+          return;
+
+        RCLCPP_ERROR(
+          self->_context->node()->get_logger(),
+          "Got ticket issueing claim");
+
+        if (msg->header.request_id != self->_reservation_id
+        || msg->header.robot_name != self->_context->name()
+        || msg->header.fleet_name != self->_context->group())
+        {
+          return;
+        }
+
+        self->_ticket = msg;
+
+        // Pick the nearest location to wait
+        auto current_location = self->_context->location();
+        if (current_location.size() == 0)
+        {
+          return;
+        }
+
+        // Order wait points by the distance from the destination.
+        std::vector<std::tuple<double, std::string, rmf_traffic::agv::Plan::Goal>>
+          waitpoints_order;
+        for (std::size_t wp_idx = 0;
+        wp_idx < self->_context->navigation_graph().num_waypoints(); wp_idx++)
+        {
+          const auto wp = self->_context->navigation_graph().get_waypoint(
+            wp_idx);
+
+          auto name = wp.name();
```

Comment:
Same as other places, here we would be pretty much copying the whole navgraph, should capture by const reference whenever possible